### PR TITLE
Exif metadata

### PR DESCRIPTION
**Pull request by faberf**
Added basic support for metadata and exif extraction. Retrieval does not work yet. It compiles and extraction seems to run properly although there is no integration test. Addresses issues #60 #59

### CODE REVIEW

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+class ExifMetadataExtractor(
+    input: Operator<Retrievable>,
+    field: Schema.Field<ContentElement<*>, MapStructDescriptor>,
+    persisting: Boolean = true
+) : AbstractExtractor<ContentElement<*>, MapStructDescriptor>(input, field, persisting, bufferSize = 1) {
+    override fun matches(retrievable: Retrievable): Boolean =
+        retrievable.filteredAttribute(SourceAttribute::class.java)?.source is FileSource /* && retrievable.type is MediaType.IMAGE */
+
+
+    fun convertType(directory: Directory, tagType: Int, type: String): Any? {
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:28]: I do not see why the type is a String, when there is a whole type system implemented in vitrivr-engine?

File: config-pipeline.json [hunk 2]
```diff
@@ -44,7 +44,16 @@
                       "nextExtractor": {
                         "fieldName": "clip",
                         "nextExtractor": {
-                          "fieldName": "dino"
+                          "fieldName": "ocr",
+                          "nextExtractor": {
+                            "fieldName": "documenttype",
```
Comment (sauterl) [inline config-pipeline.json:49]: There is no field `documenttype` on the schema, so why is this here?

File: vitrivr-engine-core/build.gradle [hunk 3]
```diff
@@ -7,6 +7,7 @@ plugins {
 dependencies {
     /** JOML dependencies for 3D mesh support. */
     implementation group: 'org.joml', name: 'joml', version: version_joml
+    implementation 'com.drewnoakes:metadata-extractor:2.19.0'
```
Comment (sauterl) [inline vitrivr-engine-core/build.gradle:10]: Please add a comment why we require this dependency / what it is for.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadata.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import org.vitrivr.engine.core.model.metamodel.Analyser
+import org.vitrivr.engine.core.model.metamodel.Schema
+import org.vitrivr.engine.core.model.query.Query
+import org.vitrivr.engine.core.model.query.bool.SimpleBooleanQuery
+import org.vitrivr.engine.core.model.retrievable.Retrievable
+import org.vitrivr.engine.core.operators.Operator
+import org.vitrivr.engine.core.operators.retrieve.Retriever
+import java.util.*
+
+private val logger: KLogger = KotlinLogging.logger {}
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadata.kt:18]: Shouldn't this be a class property for proper name resolution?

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import java.text.SimpleDateFormat
+import java.util.*
+
+class ExifMetadataExtractor(
+    input: Operator<Retrievable>,
+    field: Schema.Field<ContentElement<*>, MapStructDescriptor>,
+    persisting: Boolean = true
+) : AbstractExtractor<ContentElement<*>, MapStructDescriptor>(input, field, persisting, bufferSize = 1) {
+    override fun matches(retrievable: Retrievable): Boolean =
+        retrievable.filteredAttribute(SourceAttribute::class.java)?.source is FileSource /* && retrievable.type is MediaType.IMAGE */
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:25]: Why is the image type requirement commented? As far as I understand, EXIF is exclusively a standard for images.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt (hunk 1 above)
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:28]: Please refrain from having the type information removed here. This should definitively strictly typed and not return `Any?`. In addition, why is this nullable?

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/descriptor/struct/metadata/source/MapStructDescriptor.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    override val retrievableId: RetrievableId,
+    val columnTypes: Map<String, String>,
+    val columnValues: Map<String, Any?>,
+    override val transient: Boolean = false
+) : StructDescriptor {
+
+    companion object{
+        fun prototype(columnTypes: Map<String, String>): MapStructDescriptor {
+            val columnValues = columnTypes.mapValues { (_, type) ->
+                when (type) {
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/descriptor/struct/metadata/source/MapStructDescriptor.kt:23]: Again, I think this should build on the type system present in vitrivr-engine.

**Review by sauterl — CHANGES_REQUESTED** [review]
Thanks for the work, there are some changes required.
Also, documentation is missing, please add relevant information to the README.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt (hunk 5 above)
Comment (v0idness) [reply vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:25]: I can comment on this since this line goes back to my contribution: I intended to have an image filter at this point for the reason you mention, yet in the way it is written there it does not work and I was unsure how to write it in a manner idiomatic for vitrivr-engine.

Comment (sauterl) [reply vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:25]: This review is based on what I can access on github, hence I could not check the vitrivr-engine idomatic way of things, maybe @ppanopticon or @lucaro could enlighten us?

Comment (lucaro) [reply vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:25]: According to https://github.com/drewnoakes/metadata-extractor, the library extracts far more than just EXIF, so it would also be applicable for video, not only for images.
`retrievable.type is MediaType.IMAGE` makes no sense since `retrievable.type` is a `String`, which cannot possibly inherit from `MediaType`. You would want something like `(retrievable.filteredAttribute(SourceAttribute::class.java)?.source as FileSource).type == MediaType.IMAGE`, but then again, you'd also want to check for videos.

**Comment by faberf** [comment]
Thanks for this feedback! I will have a look at this now.

File: exif-pipeline.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    },
+    "resolver": {
+      "factory": "DiskResolver",
+      "parameters": {
+        "location": "../data"
+      }
+    }
+  },
+  "schema": "LSC-metadata",
+  "enumerator": {
```
Comment (sauterl) [inline exif-pipeline.json:14]: This seems to be pre- #53 merge. Please adjust accordingly

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import org.vitrivr.engine.core.source.MediaType
+import org.vitrivr.engine.core.source.file.FileSource
+import java.text.DateFormat
+import java.text.SimpleDateFormat
+import java.util.*
+
+
+fun parseJson(jsonString: String): Map<String, String> {
+    val map = mutableMapOf<String, String>()
+    val regex = """"([^"]*)"\s*:\s*"([^"]*)"""".toRegex()
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:25]: What is this regex for, what does it do and why is it important?

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+        for (directory in metadata.directories) {
+            for (tag in directory.tags) {
+                val tagname = tag.tagName.replace(Regex("[^a-zA-Z0-9]"), "")
+                val fullname = "${directory.name.replace(Regex("[^a-zA-Z0-9]"), "")}_${tagname}"
+
+                if (fullname == "ExifSubIFD_UserComment") {
+                    val userComment = tag.description
+                    if (userComment.isNotEmpty()) {
+                        val jsonMap = parseJson(userComment)
```
Comment (sauterl) [inline vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:80]: While I do like the approach for custom metadata do be in the user comment field, I think it is not wise to just generally assume this is JSON. I'd suggest to at least sanity check whether it _looks_ like JSON.
And otherwise treat this as textual information.

**Review by sauterl — CHANGES_REQUESTED** [review]
Thanks, I guess there is still some work to do, but some things have improved since then.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt (hunk 5 above)
Comment (lucaro) [reply vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt:64]: EXIF is indeed image only, the used library does however also support various other metadata standards, including ones for video containers.

File: vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt (hunk 9 above)
Comment (faberf) [reply vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/metadata/source/exif/ExifMetadataExtractor.kt]: I have now changed it so that if ExifSubIFD_UserComment is explicitly configured as a parameter, it will just be treated as textual information.

**Comment by faberf** [comment]
I have now addressed these concerns. Let me know if there is anything else.

**Review by sauterl — CHANGES_REQUESTED** [review]
Great! Some minor things though, otherwise this looks promising. I'll test the functionality soon ™️

**Review by sauterl — COMMENTED** [review]
Since in the meantime #55 has been merged to the `dev` branch, I'm not sure whether there are some changes required here -- @faberf could you briefly check?
Otherwise the changes made are good enough for me.

**Comment by faberf** [comment]
Merged

**Comment by faberf** [comment]
Ok if there are no more concerns, I will merge this pull request today?

**Review by sauterl — APPROVED** [review]
Looks good to me, we'll see how well it works in practice...
